### PR TITLE
[Inductor] Add xmask unswitch for dynamic pointwise kernels

### PR DESCRIPTION
**Pull request by liqiangxl**
## Summary

This adds an xmask unswitch optimization for dynamic pointwise Triton kernels.
When `xnumel` is symbolic, generated kernels normally keep `xmask` on every
load/store, even for full blocks. That prevents Triton from seeing the common
full-block path as unmasked and can block better vectorization.

With `config.triton.xmask_unswitch` enabled, kernels that have xmask-only
loads or stores now emit:

1. a full-block branch where `xmask` is replaced with a constant `True` mask
2. a tail branch that keeps the normal dynamic `xmask`

The optimization is controlled by:

```python
torch._inductor.config.triton.xmask_unswitch
```

and by the environment variable:

```bash
TORCHINDUCTOR_XMASK_UNSWITCH
```

## How It Works

`TritonKernel` first checks whether xmask unswitch may apply:

- the config flag must be enabled
- the x dimension must be dynamic

During load/store codegen, the kernel records that xmask unswitch should be
used only when the memory op is guarded solely by `xmask`.

Then `_codegen_body()` emits the usual indexing code followed by either:

- the original body for kernels without xmask-only memory ops
- an `if xoffset + XBLOCK <= xnumel` full/tail split for eligible kernels

The full-block branch reuses the existing generated load/store body, but first
sets `xmask` to a constant true mask. This avoids string-rewriting individual
load/store lines and keeps the codegen path close to the normal Triton body
emission.

## Generated Triton Code

For a simple dynamic pointwise kernel:

```python
def fn(x):
    return x * 3.0 + 1.0
```

the generated Triton code contains a full-block branch that replaces `xmask`
with a constant true mask, and a tail branch that keeps the normal dynamic
mask:

```python
@triton.jit
def triton_poi_fused_add_mul_0(
    in_ptr0, out_ptr0, xnumel, XBLOCK: tl.constexpr
):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    if xoffset + XBLOCK <= xnumel:
        xmask = tl.full([XBLOCK], True, tl.int1)[:]
        tmp0 = tl.load(in_ptr0 + (x0), xmask).to(tl.float32)
        tmp1 = tl.full([1], 3.0, tl.float32)
        tmp2 = tmp0 * tmp1
        tmp3 = tl.full([1], 1.0, tl.float32)
        tmp4 = tmp2 + tmp3
        tl.store(out_ptr0 + (x0), tmp4, xmask)
    else:
        tmp0 = tl.load(in_ptr0 + (x0), xmask).to(tl.float32)
        tmp1 = tl.full([1], 3.0, tl.float32)
        tmp2 = tmp0 * tmp1
        tmp3 = tl.full([1], 1.0, tl.float32)
        tmp4 = tmp2 + tmp3
        tl.store(out_ptr0 + (x0), tmp4, xmask)
```

## Tests

Added coverage for:

- dynamic pointwise kernels emitting the xmask split
- store-only kernels still triggering the split
- static shapes not emitting the split
- 2D kernels with both x/y masks not emitting the split
- combo kernels emitting the split inside each sub-kernel

Local validation:

```bash
lintrunner torch/_inductor/codegen/triton.py test/inductor/test_codegen_triton.py torch/_inductor/config.py
python test/inductor/test_codegen_triton.py TestXmaskUnswitch
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -1771,6 +1771,17 @@ class triton:
...
+    # Control xmask unswitch optimization for pointwise kernels with dynamic
+    # shapes. Separates full blocks (unmasked, vectorizable) from the tail
+    # block (masked) via if/else.
+    #   0: disabled (masked everywhere)
+    #   1: unswitch load/store only (compute shared between branches) — default.
+    #      Best for memory-bound kernels; avoids code bloat from duplicated compute.
+    #   2: unswitch whole body (compute duplicated in both branches).
+    #      ~10% faster on compute-heavy kernels (e.g. sin/cos) due to avoiding
+    #      phi-node overhead, but duplicates all compute in both branches.
+    xmask_unswitch: int = int(os.environ.get("TORCHINDUCTOR_XMASK_UNSWITCH", "1"))
```
Comment (eellison) [inline torch/_inductor/config.py]: I think 2 would be best as a lower level transform. we dont have a great way of doing this but we're starting to think of a restructuring that would make this possible cc @CRobeck..

File: test/inductor/test_codegen_triton.py [hunk 2]
```diff
@@ -203,6 +207,73 @@ def fn(x):
...
+    """Test the xmask_unswitch codegen optimization.
+
+    The unswitched pattern is only generated for dynamic shapes where xnumel
+    is symbolic. For static shapes, Triton already specializes on the
+    actual value via divisibility hints.
+    """
+
+    @unittest.skipUnless(HAS_GPU_AND_TRITON, "requires GPU and Triton")
+    @parametrize("mode", [1, 2])
+    def test_dynamic_shape_generates_unswitch(self, mode):
```
Comment (eellison) [inline test/inductor/test_codegen_triton.py]: can you add a test with combo kernels turned on ?

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -3813,12 +3838,13 @@ def decide_later():
...
+                if _register_xmask_load:
+                    unmasked_line += ".to(tl.float32)"
                 dtype = torch.float32
             if dtype == torch.bool and torch.version.hip is None:
-                # Workaround for https://github.com/triton-lang/triton/issues/2151
-                # tl.load returns int8 when loading from pointer to int1
-                # NOTE: Currently causes hangs on bool UTs for ROCm
                 line += ".to(tl.int1)"
+                if _register_xmask_load:
+                    unmasked_line += ".to(tl.int1)"
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: we wouldnt need to dupe this if the just emitted the switch prior

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -3813,12 +3838,13 @@ def decide_later():
                 and config.triton.codegen_upcast_to_fp32
             ):
                 line += ".to(tl.float32)"
+                if _register_xmask_load:
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: same commetn here

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -3798,6 +3816,13 @@ def decide_later():
                 shape = ()
             else:
                 line = f"tl.load({var} + ({indexing.index_str}), {indexing.mask_str}{ep}{other}{cachemod})"
+                # Register unmasked variant for xmask unswitch optimization.
+                _register_xmask_load = (
+                    self._use_xmask_unswitch
+                    and indexing.mask_vars == OrderedSet(["xmask"])
+                )
+                if _register_xmask_load:
+                    unmasked_line = f"tl.load({var} + ({indexing.index_str}), None{ep}{other}{cachemod})"
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: would it make sense to add the if/else here ?

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -6124,6 +6161,86 @@ def _has_constant_xmask(self) -> bool:
...
+    def _codegen_body_plain(self) -> None:
+        """Mode 0: plain masked code — no unswitch."""
+        for buf in [self.indexing_code, self.loads, self.compute, self.stores]:
+            self.body.splice(buf)
+
+    def _emit_unswitched_lines(
+        self, buf: IndentedBuffer, unswitch_map: dict[str, str]
+    ) -> None:
+        """Emit lines from buf, wrapping unswitchable ones in if/else."""
+        predicate = "xoffset + XBLOCK <= xnumel"
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: note: for combo kernels we generate XBLOCK1, XBLOCK2, etc. .although its possible this doesnt hit this path, im not sure.

**Review by eellison — COMMENTED** [review]
looks good,  should be basically landable a couple comments ! also, is this profitable on h100, and a100 ?

File: torch/_inductor/codegen/triton.py (hunk 6 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: Good point. A combo kernel (`ComboKernel`) fuses multiple independent sub-kernels into one Triton kernel, dispatching via `pid` ranges. Each sub-kernel is a regular `TritonKernel` instance (created with `is_combo_kernel=True`) that independently builds its own `_xmask_unswitch_map` during `load()`/`store()` codegen.
Each sub-kernel unswitches against its own `xnumel_N`, so the two features (combo dispatch and xmask unswitch) are orthogonal. The unswitch guard is nested inside the pid dispatch branch — no interaction between them. Sample code:

```python
pid = tl.program_id(0)
if pid < num_xblocks_0:                    # sub-kernel 0
    xmask = xindex < xnumel_0
    if xoffset + XBLOCK <= xnumel_0:      # unswitch: full tile
        tmp0 = tl.load(ptr, None)
    else:                                   # unswitch: partial tile
        tmp0 = tl.load(ptr, xmask)
elif pid < num_xblocks_1:                  # sub-kernel 1
    xmask = xindex < xnumel_1
    if xoffset + XBLOCK <= xnumel_1:      # same pattern, own xnumel
        ...
else:
    pass
```

File: test/inductor/test_codegen_triton.py (hunk 2 above)
Comment (liqiangxl) [reply test/inductor/test_codegen_triton.py]: Added

File: torch/_inductor/codegen/triton.py (hunk 5 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: This approach is possible, but it does not appear suitable for supporting mode-2 (whole-body unswitching). 
Mode 2 (whole-body) wraps all loads, computations, and stores inside a single if/else block. This requires visibility of all generated lines upfront, which is not available at the time of individual load() calls.

The `_xmask_unswitch_map` serves as a bridge: it provides a lightweight registration mechanism at load/store time that preserves the unmasked variant without committing to any specific control flow structure. The _codegen_body_*() methods can then consume this map after all buffers are finalized.

File: torch/_inductor/config.py (hunk 1 above)
Comment (liqiangxl) [reply torch/_inductor/config.py]: By 'lower level transform', are you refering to imperative IR?
I actually started from nvFuer's IR level unswitch, it generates explicit `how to compute` IR.  Something like: 
```
  ForLoop(i: 0..ceil(M/128))       ← explicit tiled loop                                                                              
    ForLoop(j: 0..128, Unswitch)   ← explicit inner loop with Unswitch annotation                                                     
      IfThenElse(i*128+127 < M)    ← explicit branch node                                                                             
        then: T0[i*128+j] = ...    ← no predicate (full tile)                                                                         
        else: if(i*128+j < M) T0[i*128+j] = ...  ← predicated (partial tile)
```
Then I simplified to deferred line `IfThenElse(DeferredLineBase)`:
```
then_body = IndentedBuffer()
then_body.writeline(unmasked_line)
else_body = IndentedBuffer()
else_body.writeline(masked_line)
ite = IfThenElse("xoffset + XBLOCK <= xnumel", then_body, else_body)
self.body.writeline(ite)
```

And further simplified to the current approach.

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: Avoided dupe by constructing unmasked_line after masked_line is complete.

File: torch/_inductor/codegen/triton.py [hunk 7]
```diff
@@ -3815,12 +3838,16 @@ def decide_later():
                 line += ".to(tl.float32)"
                 dtype = torch.float32
             if dtype == torch.bool and torch.version.hip is None:
-                # Workaround for https://github.com/triton-lang/triton/issues/2151
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:4357]: nit: restore comment ?

File: torch/_inductor/codegen/triton.py [hunk 8]
```diff
@@ -3815,12 +3838,16 @@ def decide_later():
...
-                # NOTE: Currently causes hangs on bool UTs for ROCm
                 line += ".to(tl.int1)"
                 dtype = torch.bool
 
+            # Build unmasked_line from final line by replacing mask with None.
+            # Done after all suffixes (.to(...)) to avoid duplicating each suffix.
+            if _register_xmask_load:
+                unmasked_line = line.replace(
+                    f", {indexing.mask_str}", ", None", 1
+                )
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: We get a bug with this pr for:

```
  import torch
  import torch._inductor.config as inductor_config

  def fn(x):
      return x.sum(dim=-1)

  x = torch.randn(4097, 1024, device="cuda")
  torch._dynamo.mark_dynamic(x, 0)

  with inductor_config.patch({"triton.xmask_unswitch": 1}):
      result = torch.compile(fn)(x)
```

Sorry, your original approach might be better here, of constructing the load string, and applying the same suffixes.

File: torch/_inductor/codegen/triton.py [hunk 9]
```diff
@@ -6124,6 +6162,86 @@ def _has_constant_xmask(self) -> bool:
...
+                self.body.writeline(f"if {predicate}:")
+                with self.body.indent():
+                    self.body.writeline(unswitch_map[stripped])
+                self.body.writeline("else:")
+                with self.body.indent():
+                    self.body.writeline(stripped)
+            else:
+                self.body.writeline(line_str)
+
+    def _codegen_body_unswitch_ldst_only(self) -> None:
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: Do you have an example of a compute bound kernel that gets faster as a result of this (mode 2) ?

Ultimately, I think for the reductions we should do something along the lines of https://github.com/pytorch/pytorch/issues/148402.

File: torch/_inductor/config.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/config.py]: This would be a lower level of IR that is post fusion, with structured control flow.

Do you have an example of a compute bound kernel that gets faster as a result of mode 2 here ?  

I think we could combine something like https://github.com/pytorch/pytorch/issues/148402 to avoid the rblock mask, and then if/else around the around the xmask.

I guess I'm asking if mode 2 is worth adding right now in relation to its complexity since it does not currently close the gap with the reduction kernels.

File: torch/_inductor/codegen/triton.py (hunk 8 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py]: sorry for churn! looking forward to this feature landed.

Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: Sorry for the delay. I was on vacation in the past week. Reverted to the previous approach.

File: torch/_inductor/codegen/triton.py (hunk 9 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: I don't have a real application case which requires mode-2. I used `y=sin(x1)+cos(x2)` during the dev.
Removed from this PR.

File: torch/_inductor/codegen/triton.py (hunk 8 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py]: no worries. also i'm at pytorch paris, will be able to get to this in a couple days.

File: torch/_inductor/codegen/triton.py (hunk 9 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py:6201]: Do we still need this, then ? could we just do the if/else aroudn the load & store?

File: torch/_inductor/codegen/triton.py (hunk 5 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py]: Since we're removing whole-body unswitching based on subsequent perf ([this comment](https://github.com/pytorch/pytorch/pull/178318#discussion_r3037108986) can we remove this ?

**Review by eellison — COMMENTED** [review]
Fix tests then i'll take another look ?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by liqiangxl** [comment]
@claude why there are still 5 CI failues?
